src/Internal/Service/DuplicatePaymentPreventionService.php

```diff
@@ -0,0 +1,237 @@
+<?php
+/**
+ * Class DuplicatePaymentPreventionService
+ *
+ * @package WooCommerce\Payments
+ */
+
+namespace WCPay\Internal\Service;
+
+use Exception;
+use WCPay\Constants\Intent_Status;
+use WCPay\Core\Server\Request\Get_Intention;
+use WCPay\Exceptions\Order_Not_Found_Exception;
+use WCPay\Internal\Proxy\HooksProxy;
+use WCPay\Logger;
+
+/**
+ * Used for methods, which detect existing payments or payment intents,
+ * and prevent creating duplicate payments.
+ */
+class DuplicatePaymentPreventionService {
+	/**
+	 * Key name for saving the current processing order_id to WC Session with the purpose
+	 * of preventing duplicate payments in a single order.
+	 *
+	 * @type string
+	 */
+	const SESSION_KEY_PROCESSING_ORDER = 'wcpay_processing_order';
+
+	/**
+	 * Flag to indicate that a previous order with the same cart content has already paid.
+	 *
+	 * @type string
+	 */
+	const FLAG_PREVIOUS_ORDER_PAID = 'wcpay_paid_for_previous_order';
+
+	/**
+	 * Flag to indicate that a previous intention attached to the order was successful.
+	 *
+	 * @type string
+	 */
+	const FLAG_PREVIOUS_SUCCESSFUL_INTENT = 'wcpay_previous_successful_intent';
+
+	/**
+	 * OrderService instance.
+	 *
+	 * @var OrderService
+	 */
+	private $order_service;
+
+	/**
+	 * HooksProxy instance.
+	 *
+	 * @var HooksProxy
+	 */
+	private $hooks_proxy;
+
+	/**
+	 * Initializes all dependencies.
+	 *
+	 * @param OrderService $order_service The order service instance.
+	 * @param HooksProxy   $hooks_proxy   The hooks proxy instance.
+	 */
+	public function __construct( OrderService $order_service, HooksProxy $hooks_proxy ) {
+		$this->order_service = $order_service;
+		$this->hooks_proxy   = $hooks_proxy;
+	}
+
+	/**
+	 * Initializes this class's hooks.
+	 *
+	 * @return void
+	 */
+	public function init_hooks(): void {
```

Comment:
Not yet register this `init_hook` in `WC_Payments` as that can make this action `clear_session_processing_order_after_landing_order_received_page` is registered twice. 

Still pending to think how to approach it.

Reply:
I'm curious where `init_hooks` of the DPPS will be called. WDYT about https://github.com/Automattic/woocommerce-payments/pull/7479 (Hooks Registry)?

I like the concept of `Hooks Registry` for new `src` folder. But it's still not getting out of how to solve potential duplicate hook registers in this specific case.

Reply:
Do you see any reason to keep the hooked methods in the legacy service? AFAICT we can remove them, and have a single hook?

Reply:
No reason to keep it! If we can guarantee that the hook is always loaded, so I think we're all good. I will dive into this idea, and see how it looks like.